lib/tree.rb

```diff
@@ -1,64 +1,170 @@
+          if !current.left.nil?
+            current = current.left
+          else
+            current.left = TreeNode.new(key, value)
+            return
+          end
+          
+        else
+          
+          if !current.right.nil?
+            current = current.right
+          else
+            current.right = TreeNode.new(key, value)
+            return
+          end
+        end 
+      end
+    end
+    
+    # if using the recursive helper method instead
+    # @root = add_helper(@root, key, value)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(log n) where n is number of nodes ?
+  # Space Complexity: O(1) or is this N/A?
   def find(key)
-    raise NotImplementedError
+    if @root.nil?
+      return nil
+    else
+      current = @root
+      while true
+        if key == current.key
+          return current.value
+        elsif key < current.key
+          current = current.left
+        else 
+          current = current.right
+        end
+      end
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+    
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+    
+    return list
+  end
+  
+  # Time Complexity: O(n) where n is number of nodes
+  # Space Complexity: O(n) where n is number of nodes
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+    
+    list << { key: current_node.key, value: current_node.value }
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+  end
+  
+  # Time Complexity: O(n) where n is number of nodes
+  # Space Complexity: O(n) where n is number of nodes 
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+    
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << { key: current_node.key, value: current_node.value }
+  end
+  
+  # Time Complexity: O(n) where n is number of nodes
+  # Space Complexity: O(n) where n is number of nodes 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  
+  
+  def height_helper(current)
+    return 0 if current.nil?
+    
+    left_height = height_helper(current.left)
+    right_height = height_helper(current.right)
+    
+    return [left_height, right_height].max + 1
+  end
+  
+  # Time Complexity: O(n) where n is number of nodes
+  # Space Complexity: O(1)
   def height
```

Comment:
Since `height_helper` is recursive, this is O(log n) for space complexity.  Since you'll have log n recursive calls on the stack at any one time.